test/forwarder/forwarder.js

```diff
@@ -105,61 +115,247 @@ contract('Forwarder', async (accounts) => {
+        return TruffleAssert.reverts(ForwarderInstance.execute(request_other, sign), "MinimalForwarder: signature does not match request");
+    });
+
+    it ('Execute should succeed even if the call to the target failed', async () => {
+        const new_domain = {
+            name,
+            version,
+            chainId: 1,
+            verifyingContract: ResponseForwarderInstance.address,
+        };
+
+        const new_request = {
+            from: relayer1Address,
+            to: ForwarderInstance.address,
+            value: '0',
+            gas: '300000',
+            nonce: 0,
+            data: '0x'
+        }
+
+        const new_sign = ethSigUtil.signTypedMessage(
+            relayer1.getPrivateKey(),
+            {
+                data: {
+                    types: types,
+                    domain: new_domain,
+                    primaryType: 'ForwardRequest',
+                    message: new_request
+                }
+            }
+        )
+
+        await ResponseForwarderInstance.response(new_request, new_sign);
+        const status = await ResponseForwarderInstance.status();
+        assert.equal(status, false);
+    });
+
+    it ('Should be failed in case of execute is called with less gas than req.gas', async () => {
+        const new_request = {
+            from: relayer3Address,
+            to: TestTargetInstance.address,
+            value: '0',
+            gas: '300000',
+            nonce: 0,
+            data: '0x'
+        }
+
+        const new_sign = ethSigUtil.signTypedMessage(
+            relayer3.getPrivateKey(),
+            {
+                data: {
+                    types: types,
+                    domain: domain,
+                    primaryType: 'ForwardRequest',
+                    message: new_request
+                }
+            }
+        )
+        const accountBalance = await provider.getBalance(accounts[0]);
+        await TruffleAssert.fails(ForwarderInstance.execute(new_request, new_sign, {value: accountBalance.toString()}));
+    });
+
+    it ('req.gas should be passed to the target contract', async () => {
+        const requestGas = 100000;
+        const new_request = {
+            from: relayer3Address,
+            to: TestTargetInstance.address,
+            value: '0',
+            gas: requestGas.toString(),
+            nonce: 0,
+            data: '0x'
+        }
+
+        const new_sign = ethSigUtil.signTypedMessage(
+            relayer3.getPrivateKey(),
+            {
+                data: {
+                    types: types,
+                    domain: domain,
+                    primaryType: 'ForwardRequest',
+                    message: new_request
+                }
+            }
+        )
+
+        await ForwarderInstance.execute(new_request, new_sign);
+        const availableGas = await TestTargetInstance.gasLeft();
+        assert(availableGas > 40000);
```

Comment:
If 100000 gas was passed into the target, then atleast 95000 should be available.